internal/runner/operation.go

```diff
@@ -42,6 +42,47 @@ func (r *Runner) executeJob(
 ) (*pb.Job_Result, error) {
 	job := assignment.Job
 
+	var task *pb.Task
+	if job.Task != nil {
```

Comment:
I was thinking we'd do all of this on the server side in the state system or the API system. I think that introduces less edge cases. When a runner acks a job, it is running, so at that point we can look up any tasks associated with the job, and update their state accordingly. We should be able to represent all state transitions cleanly this way, and much more easily test it (since state/API tests are easy unit tests to write).

In addition, I'd prevent modifying task state with `UpsertTask`, so there is no API path to break that internal state. It is purely event driven rather than trusting API clients.

Thoughts?

Reply:
This was made in a follow up PR, intended to be merged into this one! https://github.com/hashicorp/waypoint/pull/3278